cmd/xl-v1-healing.go

```diff
@@ -82,29 +86,56 @@ func (xl xlObjects) HealBucket(bucket string) error {
 	}); reducedErr != nil {
 		return toObjectErr(reducedErr, bucket)
 	}
-	return nil
+
+	// Proceed to heal bucket metadata.
+	return healBucketMetadata(xl.storageDisks, bucket)
 }
 
-// HealObject heals a given object for all its missing entries.
-// FIXME: If an object object was deleted and one disk was down, and later the disk comes back
-// up again, heal on the object should delete it.
-func (xl xlObjects) HealObject(bucket, object string) error {
-	// Verify if bucket is valid.
-	if !IsValidBucketName(bucket) {
-		return traceError(BucketNameInvalid{Bucket: bucket})
+// Heals all the metadata associated for a given bucket.
+func healBucketMetadata(storageDisks []StorageAPI, bucket string) error {
+	policyPath := pathJoin(bucketConfigPrefix, bucket, policyJSON)
+	// Lock the `policy.json` before healing.
+	policyLock := nsMutex.NewNSLock(minioMetaBucket, policyPath)
+	policyLock.RLock()
+	defer policyLock.RUnlock()
+
```

Comment:
Also, it appears that this keeps the lock longer than necessary. It may be better to write this section with this scheme:

```go
f := func(){
    policyLock := ...
    policyLock.RLock() // or write lock - not sure
    defer policyLock.RUnlock()
    return healObject(...)
}
if err := f(); err != nil {
    if !isErrObjectNotFound(err) {
        return err
    }
}
```